constellations/orion/src/repos/user_afh_repo.go

```diff
@@ -0,0 +1,169 @@
+type userAfhRepo struct {
+	db *sql.DB
+}
+
+// Interface to implement
+type UserAfhRepoInterface interface {
+	Initialize(db *sql.DB)
+	Insert(domains.UserAfh) error
+	SelectByUserId(uint) ([]domains.UserAfh, error)
+	SelectByAfhId(uint) ([]domains.UserAfh, error)
+	SelectByBothIds(uint, uint) (domains.UserAfh, error)
+	Update(uint, domains.UserAfh) error
+	Delete(uint) error
+}
+
+func (ur *userAfhRepo) Initialize(db *sql.DB) {
+	ur.db = db
+}
+
+func (ur *userAfhRepo) Insert(userAfh domains.UserAfh) error {
+	statement := "INSERT INTO user_afh (" +
+		"user_id, " +
+		"afh_id " +
+		") VALUES (?, ?)"
+
+	stmt, err := ur.db.Prepare(statement)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	execResult, err := stmt.Exec(
+		userAfh.UserId,
+		userAfh.AfhId)
+	if err != nil {
+		return err
+	}
+	return utils.HandleSqlExecResult(execResult, 1, "userAfh was not inserted")
+}
+
+func (ur *userAfhRepo) SelectByUserId(userId uint) ([]domains.UserAfh, error) {
+	results := make([]domains.UserAfh, 0)
+
+	statement := "SELECT * FROM user_afh WHERE user_id=?"
+	stmt, err := ur.db.Prepare(statement)
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+	rows, err := stmt.Query(userId)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var userAfh domains.UserAfh
+		if errScan := rows.Scan(
+			&userAfh.Id,
+			&userAfh.UserId,
+			&userAfh.AfhId); errScan != nil {
+			return results, errScan
+		}
+		results = append(results, userAfh)
+	}
+	return results, nil
+}
+
+func (ur *userAfhRepo) SelectByAfhId(afhId uint) ([]domains.UserAfh, error) {
+	results := make([]domains.UserAfh, 0)
+
+	statement := "SELECT * FROM user_afh WHERE afh_id=?"
+	stmt, err := ur.db.Prepare(statement)
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+	rows, err := stmt.Query(afhId)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var userAfh domains.UserAfh
+		if errScan := rows.Scan(
+			&userAfh.Id,
+			&userAfh.UserId,
+			&userAfh.AfhId); errScan != nil {
+			return results, errScan
+		}
+		results = append(results, userAfh)
+	}
+	return results, nil
+}
+
+func (ur *userAfhRepo) SelectByBothIds(userId, afhId uint) (domains.UserAfh, error) {
+	statement := "SELECT * FROM user_afh WHERE user_id=? AND afh_id=?"
+	stmt, err := ur.db.Prepare(statement)
+	if err != nil {
+		return domains.UserAfh{}, err
+	}
+	defer stmt.Close()
+
+	var userAfh domains.UserAfh
+	row := stmt.QueryRow(userId, afhId)
+	errScan := row.Scan(
+		&userAfh.Id,
+		&userAfh.UserId,
+		&userAfh.AfhId)
+	return userAfh, errScan
+}
+
+func (ur *userAfhRepo) Update(id uint, userAfh domains.UserAfh) error {
+	statement := "UPDATE user_afh SET " +
+		"user_id=?, " +
+		"afh_id=? " +
+		"WHERE id=?"
+	stmt, err := ur.db.Prepare(statement)
+	if err != nil {
+		return err
+	}
+	defer stmt.Close()
+
+	execResult, err := stmt.Exec(
+		userAfh.Id,
+		userAfh.UserId,
+		userAfh.AfhId,
+		id)
```

Comment:
found your problem! Looks like you're giving 4 arguments here, but there are only 3 `?`s above.

So this is where mocking is sometimes not useful. In this application, there are many moving parts. It would be very difficult to test all the moving parts at once! Instead, we want to test only certain components. But if components depend on each other (like A depends on B which depends on C...) how do we test just B? We can test B by *pretending* C works as expected. From there, the unit tests for B will test "Does B behave correctly if C works as expected?"

The _repo_test did not catch this because it's not running MySQL for real (that would require a MySQL server running and all this setup which is a hassle). The _repo_test is just making sure the correct MySQL query looks correct through pattern matching.

The _controller_test did not catch this because it depends on user_afh_repo. The _controller_test PRETENDS the repo is working correctly. If the repo works correctly, does the controller behave as expected?

So now you probably can see that there are some holes that these tests don't always catch. That's what integration tests are for and why you only bump into this problem in the integration tests.

So why not always use integration tests?
You could. But integration tests are a good chunk slower than unit tests because it requires network queries, MySQL servers & setups, etc. If it works for integration tests, it should *most likely* work correctly. However, if you had to test every single little detail with integration tests (like every possible bad inputs), the whole testing framework is going to take MUCH longer.

I should write a post in the resources folder about this lol but hopefully you have a better idea of what happened here. Super small mistake, but actually exposes a lot of nuance.

Reply:
I see. So basically before, I was modifying the id too, but it should have only been modifying the userId and afhId by the id..?

Reply:
Ah yes, what you have now is correct. There are 3 question marks in lines 128-131, so you need to make sure your arguments then line up in order. The first one is userId, second is afhId and the third is the row Id of the `userAfh` object. In this case, you were actually fine using either `id` or `userAfh.Id` because they are the same (tho I prefer the former).

There are some domains (like programs) where you can actually update the `program_id`. So there's a sense of a new programId and an old programId. When updating, you have to be careful which one you want to use.